spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/Hack.java

```diff
@@ -0,0 +1,450 @@
+				return true;
+			}
+			if (current instanceof Fuseable.ScalarCallable) {
+				return false;
+			}
+			if (isDecorator(current)) {
+				return false;
+			}
+			if (!isLifter(current)) {
+				if (!isSync(current)) {
+					return true;
+				}
+
+				if (isSourceProducer(current)) {
+					return false;
+				}
+			}
+			if (--maxSteps <= 0) {
+				return true;
+			}
+			current = getParent(current);
+		}
+	}
+
+	private static boolean isDecorator(Publisher<?> current) {
+		return current instanceof SleuthDecorator;
+	}
+
+	private static boolean isSourceProducer(Publisher<?> p) {
+		return sourceProducerClass.isInstance(p);
+	}
+
+	private static boolean isSync(Publisher<?> p) {
+		return !(p instanceof Processor) && Scannable.Attr.RunStyle.SYNC == Scannable
+				.from(p).scan(Scannable.Attr.RUN_STYLE);
+	}
+
+	@Nullable
+	private static Publisher<?> getParent(Publisher<?> publisher) {
+		Object parent = Scannable.from(publisher).scanUnsafe(Scannable.Attr.PARENT);
+		if (parent instanceof Publisher) {
+			return (Publisher<?>) parent;
+		}
+		return null;
+	}
+
+	static boolean isReactorCorePublisher(String pubClassName) {
+		return pubClassName != null && pubClassName.startsWith("reactor.core.publisher");
+	}
+
+	// should be changed to Scannable. need extra Attr or interface for Lifter to
+	// distinguish from others
+	private static boolean isLifter(Publisher<?> current) {
+		if (current == null) {
+			return false;
+		}
+		String name = current.getClass().getName();
+		return isReactorCorePublisher(name)
+				&& ((name.endsWith("Lift") || name.endsWith("LiftFuseable")));
+	}
+
+	// WA need a way to determine whether Publisher is SleuthDecorator
+	// mostly it is a copy of reactor-core logic
+	// as on option - add Scannable to Lift Publishers which returns original lifter
+	// function
+	@SuppressWarnings("unchecked")
+	public static <I, O> Function<? super Publisher<I>, ? extends Publisher<O>> liftPublisher(
+			Predicate<Publisher> filter,
+			BiFunction<Publisher, ? super CoreSubscriber<? super O>, ? extends CoreSubscriber<? super I>> lifter) {
+		Function<? super Publisher<I>, ? extends Publisher<O>> coreLifter = Operators
+				.liftPublisher(lifter);
+		return publisher -> {
+			if (filter != null && !filter.test(publisher)) {
+				return (Publisher<O>) publisher;
+			}
+
+			if (publisher instanceof Mono) {
+				return new SleuthMonoLift<>(publisher, lifter);
+			}
+			if (publisher instanceof ParallelFlux) {
+				return new SleuthParallelLift<>((ParallelFlux<I>) publisher, lifter);
+			}
+			if (publisher instanceof ConnectableFlux) {
+				return new SleuthConnectableLift<>((ConnectableFlux<I>) publisher,
+						lifter);
+			}
+			if (publisher instanceof GroupedFlux) {
+				return new SleuthGroupedLift<>((GroupedFlux<?, I>) publisher, lifter);
+			}
+			return new SleuthFluxLift<>(publisher, lifter);
+		};
+	}
+
+}
+
+final class SleuthMonoLift<I, O> extends MonoOperator<I, O> implements SleuthDecorator {
```

Comment:
As these types look like mechanical things to allow implementation of SleuthDecorator, it would be nice to add a one-line comment that "this does nothing special" or "this does XXX" to know the difference if any are doing something subtle.

Reply:
you are right, this is just a WA. Comment above is applicable to all this stuff   
> 	// WA need a way to determine whether Publisher is SleuthDecorator
> 	// mostly it is a copy of reactor-core logic
> 	// as on option - add Scannable to Lift Publishers which returns original lifter
> 	// function

I hope it would be deleted if `reactor-core` team adds access to the `lifter function`, so all this stuff would not be needed anymore.